### PR TITLE
Change the gRPC o11y logName when sending logs to GCP.

### PR DESCRIPTION
**Pull request by fengli79**
Change the gRPC o11y logName when sending logs to GCP.
From: grpc
To: microservices.googelapis.com%2Fobservability%2Fgrpc

### CODE REVIEW

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:50]: - Has this been decided?
- Need to sync this with other langs (Go)
- Do we need to percent-encode "/" ? If it works without percent-encoding I would prefer to not do it

Comment (fengli79) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:50]: - Yes, if no other objections from the community. This pull request give the visibility to the community as well.
- Yes, golang will be another PR.
- Yes, it's required to be url encoded:
image

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:205]: Could you restore the deleted newline?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java
Comment (sanjaypujare) [inline gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:51]: Couple of things about this change:

- I am surprised this change does not need any unit test change (@DNVindhya we should have a unit test to verify this log name?)
- we don't have any interop tests to verify with CloudLogging. Anyway we can verify and manually and document here?

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java
Comment (fengli79) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:205]: Resolve this conversation per our offline discussion.

File: gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java
Comment (fengli79) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:51]: - Yes, we should verify that our log entries always use this log name, checks added in our unit test.
- Run a manual test with Vindhya's help. Here's the log name shown in GCP cloud logging GUI.
- 
image

Comment (sanjaypujare) [reply gcp-observability/src/main/java/io/grpc/gcp/observability/logging/GcpLogSink.java:51]: Interesting. If the log-name appears without percent decoding (i.e. as `microservices.googleapis.com%2Fobservability%2Fgrpc` instead of `microservices.googleapis.com/observability/grpc`) then what's the use for the end user? I thought the `%2F` is only for the intermediate plumbing thru the API but in the dashboard I should expect to see a properly formatted string like `projects/grpc-testing/logs/microservices.googleapis.com/observability/grpc` . Am I missing something?

**Review by sanjaypujare — COMMENTED** [review]
Looking at the CloudLogging UI output I expect to see a properly formatted string like projects/grpc-testing/logs/microservices.googleapis.com/observability/grpc unless I am missing something?

**Comment by fengli79** [comment]
Decoding a user provided string on web GUI could be pretty dangerous.
Also, by definition, log name `projects/grpc-testing/logs/microservices.googleapis.com%2Fobservability%2Fgrpc` is a string with URL encoded log id (`microservices.googleapis.com/observability/grpc`) in the middle.
We verified this behavior by checking the log entries generated by audit log as well:
image

**Comment by sanjaypujare** [comment]
> Decoding a user provided string on web GUI could be pretty dangerous.

You are talking about things like XSS. Indiscriminately decoding user input can be bad I agree - but they also do "safe decoding" when they are sure there is no security vulnerability. Decoding `"%2F"` to `"/"` in an html output can never be dangerous.  

I have a basic question: if our (so called user provided) string of `"microservices.googleapis.com%2Fobservability%2Fgrpc"` is never rendered in a user friendly way (such as `"microservices.googleapis.com/observability/grpc"`) to the user then why even bother with a name like that? Why not just use an `_` or `-`  so  e.g. `microservices.googleapis.com_observability_grpc` ?

Also looking at the doc you cited https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry it also says

```
[LOG_ID] must be less than 512 characters long and can only include the following characters: upper and lower case alphanumeric characters, forward-slash, underscore, hyphen, and period.
```

I see some inconsistency here. They say it can contain forward-slash and I can't believe it means a percent-encoded forward-slash.

**Comment by fengli79** [comment]
%2F is not danger doesn't mean other decoding is not danger too, like %3C. Also, this is cloud logging behavior instead of gRPC behavior. Cloud logging may apply other security protection mechanism, like validate the user generated string, etc, but eventually, it's their call about how to render the UGC.

I don't see inconsistency in cloud logging's definition and its GUI about log id and log name.
Please note (I posted in previous comments as well)
- Log ID can contains forward slash `/`, and `microservices.googleapis.com/observability/grpc` is a valid Log ID.
- Log Name is a string with **URL encoded** Log ID.
- The cloud logging GUI shows the Log Name, not the Log ID.

Regarding to the `microservices.googleapis.com_observability_grpc` proposal, it's not a valid url any way (even worse for readability, also note that URL encode the log id implies its format should be a valid URL, or better to be), and we should be align with other googleapis, like audit logging. Having `microservices.googleapis.com/' prefix in our log id is important to distinguish from user generated log entries, they may not have a domain prefix and there's no clear way to distinguish them.

**Comment by sanjaypujare** [comment]
There is miscommunication. We should discuss offline to get this sorted out

**Comment by fengli79** [comment]
A few updates after our offline meeting.
1. It's a must, to set URL encoded log id with in the log name. (I forwarded the bug to you via IM)
2. Project id was set via LoggingOptions, so, no need to set it in log name. (This needs to be followed up with cloud logging, as their document is probably not accurate, it reads like required in the log name).
3. Given that log id with log name has to be URL encoded, showing a decoded log id when cloud logging GUI presents the log name is wrong, unless cloud logging change the definition of log name.

Please let me if there's other action items in your mind to unblock this PR.

**Review by sanjaypujare — APPROVED** [review]
Discussed offline with @fengli79.  We agreed that the UX is bad and the relevant backend library should be fixed to decode "%2f" to "/" and we will continue to push the owners to fix it.

For now approving it (reluctantly) to get us going

**Comment by fengli79** [comment]
@sanjaypujare, could you help to merge?

**Comment by sanjaypujare** [comment]
> @sanjaypujare, could you help to merge?

Oops, sorry. Doing it now

**Pull request merged**
